Add block completion value as optional field in course_blocks.api.

tomaszgy opened the pull request:
# Allow `completion` in `requested_fields`
This PR extends Course Blocks API to allow user to pass `completion` into `requested_fields`. API returns block level completion data for each block (0.0 - 1.0 or None if not completable).

**JIRA tickets**: Implements [OC-3114](https://tasks.opencraft.com/browse/OC-3114).

**Discussions**: This PR is a part of the implementation of the [Completion API](https://openedx.atlassian.net/wiki/spaces/OpenDev/pages/162247762/Completion+API).

**Dependencies**: It updates version `xblock` in requirements to `1.1.1`.

**Sandboxes URLs**:
LMS: https://pr16674.sandbox.opencraft.hosting
Studio: https://studio-pr16674.sandbox.opencraft.hosting/

**Merge deadline**: TBD

**Testing instructions**:
I
1. Run included tests from `lms/djangoapps/course_api/blocks/transformers/tests/test_block_completion.py` which is a part of this PR.

II
1. Go into Django shell.
2. Create entries for the completion using `BlockCompletion` as in #16047:
```
>>> from lms.djangoapps.completion.models import BlockCompletion
>>> from django.contrib.auth.models import User
>>> from opaque_keys.edx.keys import CourseKey, UsageKey
>>> BlockCompletion.objects.submit_completion(user=User.objects.get(...), course_key=CourseKey.from_string('...'), block_key=UsageKey.from_string('...'), completed=0.5)
```
3. Test that the following API call returns blocks with proper `completion` values:
http://localhost:18000/api/courses/v1/blocks/?course_id=course-v1:edX%2BDemoX%2BDemo_Course&all_blocks=true&requested_fields=completion&depth=10

(Make sure the id in the `course_id` is the same as `course_key` in point 2)

**Author notes and concerns**: None

**Reviewers**
- [x] @bradenmacdonald 
- [x] @jcdyer
- [ ] edX reviewer[s] TBD

**Settings**
As `setup.py` file has been changed to register new transformer you'll need to do `python setup.py install` inside your container.

EDIT:
Rebased on master from 1902951aeb3c0e049e221683e8dc11985102a7ee.
Squashed and then rebased from db3940d56d1736ed7f977b0584ae5a7dfec4f3f8 onto master.
Signed-off-by: Tomasz Gargas <tomasz@opencraft.com>

edx-webhook commented:
Thanks for the pull request, @tomaszgy! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=16674"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

bradenmacdonald commented on lms/djangoapps/course_api/blocks/transformers/block_completion.py:
```diff
@@ -0,0 +1,82 @@
+    BLOCK_COMPLETION = 'block_completion'
+
+    # this might be replaced with xblock.completable.CompletableXBlockMixin.{AGGREGATOR, EXCLUDED}
+    # once we change requirements/base.txt to contain xblock 1.1.1
```
Yep, please go ahead and do that now :)

tomaszgy replied:
👍

bradenmacdonald commented on lms/djangoapps/course_api/blocks/transformers/block_completion.py:
```diff
@@ -0,0 +1,82 @@
+    """
+    WRITE_VERSION = 1
+    READ_VERSION = 1
+    BLOCK_COMPLETION = 'block_completion'
```
Could we just call this "completion" ? Then it will match the request parameter, and we can use the same field for storing aggregate completions in the future once we implement that.

tomaszgy replied:
👍

bradenmacdonald commented on lms/djangoapps/course_api/blocks/transformers/block_completion.py:
```diff
@@ -0,0 +1,82 @@
+        completions_dict = dict(
+            BlockCompletion.objects.filter(
+                user=usage_info.user,
+                course_key=usage_info.course_key
```
Nit: I recommend putting trailing commas on lines like this.

bradenmacdonald commented on lms/djangoapps/course_api/blocks/transformers/tests/test_block_completion.py:
```diff
@@ -0,0 +1,119 @@
+    def test_transform_gives_value_for_completable(self):
+        course = CourseFactory.create()
+        block = ItemFactory.create(category='comp', parent=course)
+        BlockCompletion.objects.create(
```
Try using [`BlockCompletion.objects.submit_completion(...)`](https://github.com/edx/edx-platform/blob/ab13544cf27132ff01f2ab1e615790e5bc29f436/lms/djangoapps/completion/models.py#L40) instead.

tomaszgy replied:
Thanks! When I used `map_into_course` in the `transform()` method as you suggested in OC-3114, the problem is solved.

bradenmacdonald replied:
@tomaszgy Great - but even still, I think we should still be using `submit_completion(...)` here, so that BlockCompletion objects get created consistently using that method everywhere in the codebase. You may also need to use [this decorator](https://github.com/edx/edx-platform/blob/c1e878fb778ba96567fd54d6a84f27ce3cdab5fc/lms/djangoapps/completion/test_utils.py#L12) on this test method, to enable the completion API.

tomaszgy replied:
👍

bradenmacdonald commented:
@tomaszgy You've still got a quality issue:

```
23:42:12 lms/djangoapps/course_api/blocks/transformers/tests/test_block_completion.py (99.1%):
23:42:12     107: C0330: (bad-continuation), : Wrong hanging indentation before block.
```

bradenmacdonald commented on lms/djangoapps/course_api/blocks/transformers/blocks_api.py:
```diff
@@ -59,6 +62,7 @@ def transform(self, usage_info, block_structure):
         """
         Mutates block_structure based on the given usage_info.
         """
+        BlockCompletionTransformer().transform(usage_info, block_structure)
```
@tomaszgy  Adding this new transformer to `BlocksAPITransformer` causes it to generate its data every time a course's block structure is created, which results in one additional MySQL query per course when using the course blocks API. That is why a number of tests are failing.

Since we generally won't need to read this data and it doesn't affect block visibility or anything, I think it would be better to only execute this transformer when its field has been explicitly requested, which would involve adding logic [here](https://github.com/edx/edx-platform/blob/9c4869c1d59899a1e4a0ed2bf9cf92c77acc4447/lms/djangoapps/course_api/blocks/api.py#L52-L67) to only add this new transformer when `completion` is in `requested_fields` (or pass `requested_fields` to this class's init method and keep the logic here).

@nasthagiri does that make sense?

nasthagiri replied:
+1

jcdyer commented:
jenkins run quality

jcdyer commented on lms/djangoapps/course_api/blocks/api.py:
```diff
@@ -52,6 +53,7 @@ def get_blocks(
     include_special_exams = False
+    include_completion = requested_fields is not None and 'completion' in requested_fields
     if requested_fields is not None and 'special_exam_info' in requested_fields:
         include_special_exams = True
```
Nit: There are multiple lines here checking if requested_fields is not None.  Also, include_completion is injected into the middle of the logic for checking special_exams (L54 & LL57-58).  Consider changing LL54-58 to

    if requested_fields is None:
        requested_fields = []
    include_completion = 'completion' in requested_fields
    include_special_exams = 'special_exam_info' in requested_fields

tomaszgy replied:
👍

jcdyer commented on lms/djangoapps/course_api/blocks/transformers/block_completion.py:
```diff
@@ -0,0 +1,86 @@
+    COMPLETION = 'completion'
+
+    def __init__(self):
+        pass
```
Are you deliberately trying to suppress the behavior of the superclass's `__init__()` method, or is this just vestigial code?   This could either use an explanatory comment, or should be deleted.

tomaszgy replied:
Well spotted :-) Actually superclass doesn't define `__init__` and inherits itself from `object` but I'll add a call to superclass's constructor.

jcdyer replied:
If you're not doing anything else, just delete the definition altogether.  If you leave it out, it will search the superclasses automatically.

jcdyer commented on lms/djangoapps/course_api/blocks/views.py:
```diff
@@ -128,6 +128,10 @@ class BlocksView(DeveloperErrorViewMixin, ListAPIView):
+          * completion: (float or None) The level of completion of the block.
+            Its value can vary between 0.0 and 1.0 or be equal to None
+            if block is not completable.
+
```
This should mention that the value is only included if "completion" included in requested_fields.  See `graded` or `format` below.

tomaszgy replied:
👍

nasthagiri commented:
Hey there!  Thanks for these changes.  Overall, they look great.

When you have a moment, please read the [Transformer Version Updates](https://openedx.atlassian.net/wiki/spaces/EDUCATOR/pages/154861855/Block+Structure+Cache+Invalidation+Proposal#BlockStructureCacheInvalidationProposal-TransformerVersionUpdates
) section of the wiki. 

Essentially, if you can merge these changes in 2 phases: initially, setting the `WRITE_VERSION=1` and then following up with setting the `READ_VERSION=1` - it will allow for smoother releases into environments (such as edx.org) with blue-green deployments.  FYI @edx/devops 

Also, you'll want to update [setup.py](https://github.com/edx/edx-platform/blob/5dec436f6bd50e7250329915391c3a4c4681812d/setup.py#L51) to include the new Transformer.

nasthagiri commented on lms/djangoapps/course_api/blocks/transformers/__init__.py:
```diff
@@ -63,5 +64,10 @@ def __init__(
+    SupportedFieldType(
+        BlockCompletionTransformer.COMPLETION,
+        BlockCompletionTransformer,
+        'completion'
```
*Future note to self (and others):* It would be great if we can make this pluggable, so new Transformers don't need to manually update this list - and can be automatically extracted from registered transformers.

jcdyer replied:
I remember implementing a registry for something once, and I found two basic ways of doing it: The first is by creating a registry class with a `register()` decorator function:

    @transformers.register
    class MyTransformer(BlockTransformer):

The other uses a metaclass defined on the superclass.  Then when the class gets constructed, it is automatically registered in the metaclass.  Django models use this method.

The major differences are:

1.  that the decorator is visible to the programmer, while the metaclass is hidden.
2.  programmers can opt-out of the decorator by not including it, but not the metaclass (unless the metaclass explicitly provides a hook, like writing):
        class Registerable(Registry):
            register = False
    The downside of allowing programmers to opt out (or more accurately opt not to opt-in) is that you also make it possible to *forget* to opt-in.
3.  That decorators are easier to read and write than metaclasses.

nasthagiri commented on lms/djangoapps/course_api/blocks/api.py:
```diff
@@ -52,6 +53,7 @@ def get_blocks(
     # create ordered list of transformers, adding BlocksAPITransformer at end.
     transformers = BlockStructureTransformers()
     include_special_exams = False
+    include_completion = requested_fields is not None and 'completion' in requested_fields
```
*Future note to self (and others):* If there are any great ideas of making this automatically pluggable as well, welcoming ideas/contributions.

tomaszgy commented:
@jcdyer and @nasthagiri - thanks for your reviews! I've just addressed your comments. Please have a look and if all's good I'll squash the commits.

@jcdyer, thanks for spotting `TransformerException` - it was due to my transformer not being registered in `setup.py` as @nasthagiri pointed out. I should have spotted it before but in its initial version my transformer was living inside `BlocksAPITransformer` which has already been registered.

@nasthagiri, I've removed for now `READ_VERSION=1`, leaving just `WRITE_VERSION=1`. Do I understand correctly than I should open another PR or re-open this one, more than 24h after this PR is merged?

nasthagiri commented:
Thanks @tomaszgy 
I double checked with our devops team here.  We'd like to follow the following process for now:
1. First PR: Merge your PR with `READ_VERSION=0` and `WRITE_VERSION=1`
2. Second PR: Create and be ready to merge your 2nd PR with `READ_VERSION=1`
3. Wait until you see a message on the First PR that says:
```
EdX Release Notice: This PR has been deployed to the production environment.
```
4. Merge your Second PR

We hope that works for your team.  Thanks.

jcdyer commented:
@tomaszgy There are a bunch of test failures.  Do you know what's causing them?

tomaszgy commented:
@nasthagiri, thanks for the scenario!

@jcdyer, indeed, there're lots of tests failing. Many of them (e.g. [this one](https://build.testeng.edx.org/job/edx-platform-python-unittests-pr/45478/testReport/junit/openedx.core.djangoapps.content.block_structure.tests.test_signals/CourseBlocksSignalTest/test_course_update/)) fail due to `TransformerException('Version attributes are not set on transformer {0}.', transformer.name())` which is raised [here](https://github.com/edx/edx-platform/blob/bb9e5922e12c805668dc09697ccfa8623ec1d049/openedx/core/djangoapps/content/block_structure/block_structure.py#L733) and is an effect of `READ_VERSION=0`. Would you have any suggestions on that, @nasthagiri?

nasthagiri commented:
@tomaszgy thanks for bringing up the exception issue.  Can you change that particular block_structure code to remove the check for READ_VERSION=0?  You are running into it since you are the first to introduce a new transformer since we designed the cache-invalidation strategy.   We used that strategy for updating existing transformers, but not (yet) for introducing a new one.  Thanks.

tomaszgy commented:
Thanks for the suggestion, @nasthagiri! I've just removed the check.

tomaszgy commented:
@nasthagiri one of the bokchoy tests (`acceptance.tests.lms.test_teams.BrowseTopicsTest.test_sort_topics_2___team_count___True_ `) failed for this PR with:
`Timed out waiting to load page '<common.test.acceptance.pages.lms.teams.BrowseTopicsPage object at 0x7fba5a549490>' at URL 'http://localhost:8003/courses/course-v1:test_org+306648626320279492720022822088892457631+test_run/teams/#browse'`

In the meantime I opened Second PR (#16859) with `READ_VERSION=1` and all tests passed there. Shall we investigate why it fails here?

CC @jcdyer

mtyaka commented:
jenkins run bokchoy

tomaszgy commented:
Thanks @mtyaka! :)

tomaszgy commented:
@nasthagiri Just a status update: I've squashed and rebased this PR, now bokchoy tests pass, too. If there's anything else I should change, please let me know. If all's good, please feel free to merge it :) 

Second PR with `READ_VERSION=1` is in #16859.

nasthagiri commented:
@tomaszgy Sounds good.  Merging now.
FYI @edx/devops We'll want to run the generate_course_blocks management command AFTER this PR goes to stage and prod and BEFORE #16859 does.  I'll create a devops ticket for this.

tomaszgy commented:
@nasthagiri, thanks! Just rebased Second PR to make it ready for merge.

nasthagiri commented:
Devops ticket created: https://openedx.atlassian.net/browse/DEVOPS-6904

nasthagiri commented:
Updating here: We've run the generate_course_blocks command on Stage.  I've verified there are no exceptions raised by the new transformer.

nasthagiri commented:
Update: The management command was run on production after yesterday's release.  We're good to move forward with the next PR.
